### ISSUE TITLE
HTML export - rendering issues

### ISSUE THREAD

**Issue by eduardobbs** [archimatetool/archi#679]
### Version of Archi
4.7.1

### Archi Plug-ins
None

### Operating System
Windows Server 2019 / Windows 10 

### Expected Behaviour
Given a view which uses a Wingdings 2 font
When I generate an HTML export
Then I would expect it to be correctly rendered

### Actual Behaviour
The characters are replaced by other ones during the export

### Steps to Reproduce the Behaviour
1. On a view, create a note and set the font to Wingdings 2 
2. Type "w" (or any other key) inside the note and check the equivalent Wingding 2 character/symbol
3. Export to HTML and check the difference

Interesting notes:
The other exports ("Copy As Image to Clipboard" or "Export --> View as Image") render it correctly. Since the HTML export also creates images of the views, maybe it should reuse the same mechanism. There are also other rendering problems with the "Customer" element and the border colors in the attached images. They are not related to Wingdings 2, but would also benefit from reusing the same image export mechanism as I have just mentioned.
**Wingings x Wingdings 2**: Wingdings work just fine, the problem is really just with Wingdings2

![image](https://user-images.githubusercontent.com/5373152/95316227-7321bf80-088b-11eb-8326-92d2c6aceabc.png)
![image](https://user-images.githubusercontent.com/5373152/95316247-787f0a00-088b-11eb-9905-0a8013425839.png)

**Comment by Phillipus** [comment]
I tested this, and it works for me.

> Since the HTML export also creates images of the views, maybe it should reuse the same mechanism. 

It use exactly the same mechanism.

> There are also other rendering problems with the "Customer" element and the border colors in the attached images. They are not related to Wingdings 2, but would also benefit from reusing the same image export mechanism as I have just mentioned.

Strange. As I just mentioned it uses exactly the same mechanism.

Try on a different machine and see if it is reproducible. There should be no difference when creating image files as it's the same code.

**Comment by eduardobbs** [comment]
I think you are right.
The export is being done by an Azure Devops build agent, not on my machine.
What kind of things could be impacting it? Maybe the server does not have the Wingdings 2 font, but then what about the border colors, it is proably something on that Archi instance setting, I will check both things.

**Comment by Phillipus** [comment]
> The export is being done by an Azure Devops build agent, not on my machine.

The problem with rendering images using the Eclipse framework is the Eclipse `org.eclipse.swt.Image` class. This renders an image to the current system `Device`. The system Device is the current graphics driver (physical or virtual). So, in your case, this will be the graphics  device used by the build agent. So there may be issues with that.

Ideally, when generating an Image we would like to specify something like our own  _internal_ virtual device for rendering so that we could get the same result each time. Unfortunately, Eclipse only uses system devices, either physical or virtual.

**Comment by jbsarrodie** [comment]
> The problem with rendering images using the Eclipse framework[...]

In this case the issue is (IMHO) only this: as soon as you generate the report on another machine, you absolutely need to have the exact same set of fonts installed and the exact same Archi's configuration.

>Maybe the server does not have the Wingdings 2 font,

It clearly don't have

>what about the border colors

Open Archi on this server and you'll most certainly see that "Derive element line colours from fill color" is set (this is the default):
![Screenshot from 2020-10-07 20-10-34](https://user-images.githubusercontent.com/5757396/95370572-44323a80-08d9-11eb-9709-66446a070e43.png)

**Comment by eduardobbs** [comment]
> > The problem with rendering images using the Eclipse framework[...]
> 
> In this case the issue is (IMHO) only this: as soon as you generate the report on another machine, you absolutely need to have the exact same set of fonts installed and the exact same Archi's configuration.
> 
> > Maybe the server does not have the Wingdings 2 font,
> 
> It clearly don't have
> 
> > what about the border colors
> 
> Open Archi on this server and you'll most certainly see that "Derive element line colours from fill color" is set (this is the default):
> ![Screenshot from 2020-10-07 20-10-34](https://user-images.githubusercontent.com/5757396/95370572-44323a80-08d9-11eb-9709-66446a070e43.png)

I thought the same. Unfortunately however, I have just installed the font (wasn't there) and set the "Derive element line colours from fill color" and still the same result :(

**Comment by Phillipus** [comment]
> In this case the issue is (IMHO) only this: as soon as you generate the report on another machine, you absolutely need to have the exact same set of fonts installed and the exact same Archi's configuration.

> Open Archi on this server and you'll most certainly see that "Derive element line colours from fill color" is set (this is the default

Good call.

So there are two things:

1. Font rendering will be different (see https://forum.archimatetool.com/index.php?topic=918.0)
2, The only way a border will be different will be as JB says.

Try changing the line colors again.

**Comment by eduardobbs** [comment]
> 1. Font rendering will be different (see https://forum.archimatetool.com/index.php?topic=918.0)
I haven't finished reading this, but I am only working with Windows.

I have just tested a few things from within the server, always using the same portable instance (4.7.1). Let me explain what I tried so far:
1-Tried using the graphical UI, running under my user account (which is admin of the server)
Result: Everything worked perfectly

2- Using the CLI with my user account
Result: Wingdings 2 working fine, but border colors not

3- Using the CLI with the same account used by the build agent (NT authority\Network Service)
Result: same as number 2 above

**Comment by Phillipus** [comment]
It has to be the case that the preferences are not being picked up. Try copying them across.

**Comment by eduardobbs** [comment]
Sorry, what do you mean by copying?
Because the different test results are all using the same instance, therefore it should be using the same ini folder mappings and ,consequently, the same preferences too.

**Comment by Phillipus** [comment]
My misunderstanding. I thought they were different.

But it seems that somehow the preference is not being picked up.

The preference file is `.metadata\.plugins\org.eclipse.core.runtime\.settings\com.archimatetool.editor.prefs`

And the setting inside that file should be:

`deriveElementLineColor=false`

If that setting is not present then its value is true.

**Comment by eduardobbs** [comment]
The setting is present, I can confirm that:
![image](https://user-images.githubusercontent.com/5373152/95520499-538bb380-09bf-11eb-9799-f67f7d25918a.png)

**Comment by Phillipus** [comment]
This looks wrong:

`-Dosgi.instance.area=instance`

Perhaps you meant:

`-Dosgi.instance.area=@user.home/AppData/Roaming/ArchiPortable`

**Comment by eduardobbs** [comment]
My intention with this setting is to create everything under the same portable folder, instead of having Archi files in different places.
It seems that Dosgi.configuration.area does not work properly, this is what happens when I try to set it to "C:\ArchiPortableConfig":
![image](https://user-images.githubusercontent.com/5373152/95578777-5ec8f880-0a2c-11eb-9e75-0aacb86bd2b7.png)
It tries to create some files on the root folder, but the user account does not have write access to it, hence the error.
It uses relative path to write to a folder one level below. This became more evident when I set it to a second level folder and executed the CLI again:
![image](https://user-images.githubusercontent.com/5373152/95579062-d4cd5f80-0a2c-11eb-8c35-2aa92cafdf3b.png)
You can see the artifacts.xml and p2 folders which were not supposed to be there.

**Comment by Phillipus** [comment]
Perhaps the best thing is to set these locations to something that works.

**Comment by jbsarrodie** [comment]
Maybe be best thing is to first create a fully portable version of Archi following this [wiki page](https://github.com/archimatetool/archi/wiki/How-to-create-a-packaged-version-of-Archi-(including-configuration-and-plugins)) and especially [this part](https://github.com/archimatetool/archi/wiki/How-to-create-a-packaged-version-of-Archi-(including-configuration-and-plugins)#custom-path-for-archi4-user-profile-folder).

When you're sure this setup is working for you and uses a known path for configuration which is not related to a user profile, then running this instance from command line should work.

**Comment by Phillipus** [comment]
> It seems that Dosgi.configuration.area does not work properly, this is what happens when I try to set it to "C:\ArchiPortableConfig"

**Do not set this this to be the same folder as Archi.** It will mess with the "p2" folder and other folders. Also, when Archi starts (in non-CLI mode) it cleans up the configuration and deletes certain files in order for Archi plug-ins to work.

**Comment by Phillipus** [comment]
So here's how I spent my Sunday...

Because of potential conflicts with the Archi installation "p2" folder and the one that Eclipse generates in the parent folder of the `osgi.configuration.area` I've added a new entry to Archi.ini:

`-Declipse.p2.data.area=@config.dir/.p2`

This puts the ".p2" folder inside of the `osgi.configuration.area` and also adds a "." to "p2" just in case the user sets the `osgi.configuration.area` to the same location as the Archi location folder.

I've also changed the "cleanup" code to be safer when deleting files.

This won't solve the OP's immediate problem but that's caused by a preference file not being picked up.

**Comment by Phillipus** [comment]
Getting back to the OP...

I installed portable Archi with this `Archi.ini` file:

```
-clean
--launcher.defaultAction
openFile
-vmargs
-Dosgi.requiredJavaVersion=1.8
-Dosgi.instance.area=instance
-Dosgi.configuration.area=instance/config
-Dorg.eclipse.equinox.p2.reconciler.dropins.directory=%user.home%/AppData/Roaming/Archi4/dropins
--add-modules=ALL-SYSTEM
```

I ran Archi in UI mode and set the preference for "Derive element line colours from fill color" to false.

I ran Archi from the CLI:

```
Archi -application com.archimatetool.commandline.app -consoleLog -nosplash
--loadModel "C:\\test.archimate" --html.createReport "C:\\OutputFolder"
```

And the element borders were as expected - coloured.

**Comment by eduardobbs** [comment]
With your help I have managed to solve the problems now. 

TLDR:
**Wingdings2**: I installed the font for all users on the server, including the _NT\Network Service_ user
**Border colors / preferences**: I used psexec to run Archi UI as _NT\Network Service_ and set the preferences accordingly.

Some explanation:
Windows OS deals with these built-in accounts differently, in terms of user profile, and I struggled a bit with that. These accounts have a different profile folder path in the OS, for instance.
When using a regular user account, those ini settings worked fine (with _instance_ and _dropins_ in the same subfolder, _config_ in the user profile). 
However, for some reason, it did not work "properly" when using _Network Service_. I mean, the user preferences I copied across were not picked up, unless I manually opened the UI and set it up "again".

If you want more info to try to reproduce it just let me know. But it is working now, and maybe the thing is that these special built-in accounts were not meant to behave as regular user accounts and depend on user profile things.

I appreciate the attention given to this issue, it was very helpful. 
Archi is awesome and you have been doing a great job.

**Comment by Phillipus** [comment]
> Wingdings2: I installed the font for all users on the server, including the NT\Network Service user

Yes, I had forgotten that a font needs to be installed on the system.

**Comment by Phillipus** [comment]
Perhaps the best thing is to use relative or absolute paths for the `osgi.configuration.area` and instance locations, rather than using `@user.home`.

The next version of Archi makes it possible to have `osgi.configuration.area` as a direct sub-directory of the installation dir.  


I'll close this one one.
